src/os/ObjectStore.h

```diff
@@ -1718,22 +1718,127 @@ class ObjectStore {
-     return read(c->get_cid(), oid, offset, len, bl, op_flags);
-   }
+    uint32_t op_flags = 0) {
+    return read(c->get_cid(), oid, offset, len, bl, op_flags);
+  }
+
+  /*********************************
+   * ReadTransaction
+   *
+   * A Transaction represents a sequence of primitive non-mutating
+   * operations over the same ghobject in the same collection.
+   */
+  class ReadTransaction {
+  public:
+    virtual ~ReadTransaction() = default;
+
+    /**
+     * read -- read a byte range of data from an object.
+     *
+     * The operation can be performed synchronously or asynchronously
+     * solely depnding on the implementation's decision.
+     *
+     * Note: if reading from an offset past the end of the object, we
+     * return 0 (not, say, -EINVAL).
+     *
+     * @param offset location offset of first byte to be read
+     * @param len number of bytes to be read
+     * @param op_flags is CEPH_OSD_OP_FLAG_*
+     * @param destbl output bufferlist
+     * @param on_complete callback to be called
+     * @returns number of bytes read on success,
+     *          or negative error code on failure,
+     *          or -EINPROGESS if implementations decided to go async.
+     */
+    virtual int read(
+      uint64_t offset,
+      uint64_t length,
+      uint32_t flags,
+      ceph::bufferlist& destbl,
+      Context* on_complete) = 0;
+
+    /**
+     * read_sparse -- read a byte range of data from an object using
+     * the formatting of CEPH_OSD_OP_SPARSE_READ for output.
+     *
+     * The operation can be performed synchronously or asynchronously
+     * solely depnding on the implementation's decision.
+     *
+     * Note: if reading from an offset past the end of the object, we
+     * return 0 (not, say, -EINVAL).
+     *
+     * @param offset location offset of first byte to be read
+     * @param len number of bytes to be read
+     * @param op_flags is CEPH_OSD_OP_FLAG_*
+     * @param destbl output bufferlist
+     * @param on_complete callback to be called
+     * @returns number of bytes read on success,
+     *          or negative error code on failure,
+     *          or -EINPROGESS if implementations decided to go async.
+     */
+    virtual int read_sparse(
+      uint64_t offset,
+      uint64_t length,
+      uint32_t flags,
+      ceph::bufferlist& destbl,
+      Context* on_complete) = 0;
+
+    virtual bool empty() const = 0;
+    virtual int apply(Context* on_complete) = 0;
+  };
+
+  /**
+   * create ReadTransaction -- instantiate concrete implementation
+   * of ReadTransaction.
+   *
+   * @param cid collection for object
+   * @param oid oid of object
+   * @returns valid pointer of nullptr if an implementation doesn't
+   *          support the new ReadTransaction subinterface.
+   */
+  virtual std::unique_ptr<ReadTransaction> create_read_transaction(
+   const CollectionHandle& c,
+   const ghobject_t& oid) {
+    return nullptr;
+  }
+
+  virtual bool has_async_read() const {
```

Comment:
Drop this after adding fallback, synchronous implementation of `ReadTransaction` for other `ObjectStores`. This would allow for significant clean-up in `PrimaryLogPG` as the async path (`is_erasure() == true`) could be unified with the sync one.